character_map.h

```diff
@@ -0,0 +1,80 @@
+/*
+ * Copyright:  Copyright Dimakopoulos Theodoros 2019
+ * License:    $(LINK2 http://www.boost.org/LICENSE_1_0.txt, Boost License 1.0)
+ * Authors:    Dimakopoulos Theodoros
+ */
+#pragma once
+
+#include <cctype>
+using std::toupper;
+/*
+The character-properties map array.
+ch_map[c] is the properties of the character c.
+The bits of ch_map[c] correspond to it's properties.
+
+ch_map[c] & CM_DIL_CHAR
+will be true if c is a DIL character (first bit ON).
+
+ch_map[c] == CM_DIL_CHAR
+will be true if only that (only first bit ON).
+
+ch_map[c] == CM_DIL_CHAR | CM_DECIMAL
+will be true if c is DIL and a decimal digit
+(only therse two bits). To check for a decimal prefer 
+
+ch_map[c] & CM_DECIMAL
+because for example 4 is both octal and decimal,
+so it's  CM_DIL_CHAR | CM_DECIMAL | CM_OCTAL
+*/
+// CM for 'C'haracter 'M'ap. This way we avoid namespaces.
+enum CM_TYPE {
+  CM_NO_PROPERTY = 0,
+  CM_DIL_CHAR = 1,        // character allowed in DIL
+  CM_BLANK = 1 << 2,      // space, new-line or tab
+  CM_IDENTIFIER = 1 << 3, // if could be in an ident
+  CM_OCTAL = 1 << 4,      // '0' to '7'
+  CM_DECIMAL = 1 << 5,    // '0' to '9'
+  CM_HEXADECIMAL = 1 << 6 // '0'-'9', 'a'-'f', 'A'-'F'
+};
+
+int ch_map[128] = {}; //     D E F I N I T I O N

```

Comment:
No need for this comment.

Reply:
Btw, the rest of the code works only if you assume that every element in this array is 0. This _is_ the case, but do you know why?

Reply:
Yep. I did it on purpose. It is the ={};

Reply:
Great, but, you could remove that and still be sure that it will be initialized to 0. Why?

Reply:
No idea. I have tried such things before in c++ and arrays don't initialize automatically. Will it be?

Reply:
The short answer is: Globals are implicitly initialized to 0 if they have no explicit initialization. Generally, this will do it, but:

You can find this in the [C++ standard](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2017/n4713.pdf), paragraph `6.8.3.2 Static initialization`, point 2:
> If constant initialization is not performed, a variable with static storage duration (6.6.4.1) or thread storage duration (6.6.4.2) is zero-initialized.

(We're in the storage duration case, because it is a global but static global duration also have e.g. the `static` variables in a function).

Of course, nobody expects you or me to know the standard inside out, we're not language lawyers. But it's a good reminder that sometimes the standard is useful if you want to be super specific about something.